[quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`

jerryzh168 opened the pull request:
Stack from [ghstack](https://github.com/ezyang/ghstack):
* #29416 [quant] Make PerChannelMinMaxObserver scriptable using `torch.jit.ignore`
* **#27950 [quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`**

Summary:
att

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D18360139](https://our.internmc.facebook.com/intern/diff/D18360139)

jerryzh168 committed:
[quant][graphmode] Accept observers with ignored functions in graph mode

Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make observers scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Accept observers with ignored functions in graph mode"


Summary:
Make HistogramObserver scriptable with `@torch.jit.ignore`

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`"


Summary:
att

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 commented:
This is ready for review please take a look

ZolotukhinM commented on test/test_quantization.py:
```diff
@@ -696,6 +698,39 @@ def test_single_liner(self):
+        result_script = model_script._c._get_method('forward')(self.calib_data[0][0])
+        self.assertEqual(result_eager, result_script)
+
+    @_tmp_donotuse_dont_inline_everything
```
The default behavior has recently changed to not inline everything by default - please verify that we can remove this decorator.

jerryzh168 replied:
seems to be working

ZolotukhinM commented on torch/quantization/observer.py:
```diff
@@ -298,7 +299,10 @@ def __init__(self, dtype=torch.quint8, qscheme=torch.per_tensor_affine,
     def forward(self, x_orig):
         r"""Records the running minimum and maximum of ``x``."""
-        x = x_orig.detach()
+        return self._forward(x_orig)
```
Why do we need this?

jerryzh168 replied:
oh, we probably don't need this, I was trying to ignore _forward previously, but it turns out that it doesn't work.

ZolotukhinM commented on torch/csrc/jit/script/module.cpp:
```diff
@@ -244,7 +244,7 @@ Module Module::clone_impl(
       r.register_attribute(
           s.name,
-          s.value.type(),
+          type()->getAttribute(s.name),
```
This is an unrelated change.

ZolotukhinM commented on torch/quantization/observer.py:
```diff
@@ -700,9 +704,11 @@ def _non_linear_param_search(self):
     def _combine_histograms(
         self, dst_histogram, dst_min, dst_max, src_histogram, src_min, src_max
     ):
+        # type: (Tensor, float, float, Tensor, float, float) -> Tensor
```
Do you need this if the function is ignored?

jerryzh168 replied:
yeah, we need this otherwise everything will be treated as Tensor

ZolotukhinM commented on torch/quantization/observer.py:
```diff
@@ -750,46 +756,50 @@ def _combine_histograms(
             # remaining should go to dst_bin2
             if dst_bin_cnt < src_bin_count:
                 dst_histogram[dst_bin2] += src_bin_count - dst_bin_cnt
+        return dst_histogram
```
This seems to be an unnecessary change.

jerryzh168 replied:
this is needed for compile as well, ignored functions seems to take Tensor as return only

ZolotukhinM commented on test/test_quantization.py:
```diff
@@ -696,6 +698,39 @@ def test_single_liner(self):
             [self.calib_data],
             inplace=False)
         result_eager = model_eager(self.calib_data[0][0])
+        torch._C._jit_pass_quant_fusion(model_script._c._get_module('fc1')._get_method('forward').graph)
```
Where does this come from?

jerryzh168 replied:
...ntot sure, will remove

ZolotukhinM reviewed (approved):
Looks good, but please remove unnecessary changes and split out unrelated changes and avoid that in future.

jerryzh168 committed:
Update on "[quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`"


Summary:
att

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`"


Summary:
att

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D18360139](https://our.internmc.facebook.com/intern/diff/D18360139)

[ghstack-poisoned]

jerryzh168 committed:
Update on "[quant][graphmode] Make HistogramObserver scriptable with `@torch.jit.ignore`"


Summary:
att

Test Plan:
python test/test_quantization.py

Reviewers:
pt1quant

Subscribers:

Tasks:

Tags:

Differential Revision: [D18360139](https://our.internmc.facebook.com/intern/diff/D18360139)

[ghstack-poisoned]
